core/src/test/scala/integration/kafka/api/TransactionsTest.scala

```diff
@@ -0,0 +1,314 @@
+    TestUtils.createBrokerConfigs(numServers, zkConnect, true).map(KafkaConfig.fromProps(_, serverProps()))
+  }
+
+  @Before
+  override def setUp(): Unit = {
+    super.setUp()
+    val numPartitions = 3
+    val topicConfig = new Properties();
+    topicConfig.put(KafkaConfig.MinInSyncReplicasProp, 2.toString)
+    TestUtils.createTopic(zkUtils, topic1, numPartitions, numServers, servers, topicConfig)
+    TestUtils.createTopic(zkUtils, topic2, numPartitions, numServers, servers, topicConfig)
+  }
+
+  @After
+  override def tearDown(): Unit = {
+    super.tearDown()
+  }
+
+  @Test
+  def testBasicTransactions() = {
+    val producer = transactionalProducer("my-hello-world-transactional-id")
+    val consumer = transactionalConsumer()
+    try {
+      producer.initTransactions()
+
+      producer.beginTransaction()
+      producer.send(producerRecord(topic1, "1", "1", willBeCommitted = true))
+      producer.send(producerRecord(topic2, "3", "3", willBeCommitted = true))
+      producer.commitTransaction()
+
+      producer.beginTransaction()
+      producer.send(producerRecord(topic2, "2", "2", willBeCommitted = false))
+      producer.send(producerRecord(topic1, "4", "4", willBeCommitted = false))
+      producer.abortTransaction()
+
+      consumer.subscribe(List(topic1, topic2))
+
+      val records = pollUntilExactlyNumRecords(consumer, 2)
+      records.zipWithIndex.foreach { case (record, i) =>
+        assertCommittedAndGetValue(record)
+      }
+    } catch {
+      case e @ (_ : KafkaException | _ : ProducerFencedException) =>
+        fail("Did not expect exception", e)
+    } finally {
+      consumer.close()
+      producer.close()
+    }
+  }
+
+  @Ignore @Test
+  def testSendOffsets() = {
+    // The basic plan for the test is as follows:
+    //  1. Seed topic1 with 1000 unique, numbered, messages.
+    //  2. Run a consume/process/produce loop to transactionally copy messages from topic1 to topic2 and commit
+    //     offsets as part of the transaction.
+    //  3. Randomly abort transactions in step2.
+    //  4. Validate that we have 1000 unique committed messages in topic2. If the offsets were committed properly with the
+    //     transactions, we should not have any duplicates or missing messages since we should process in the input
+    //     messages exactly once.
+
+    val transactionalId = "foobar-id"
+    val consumerGroupId = "foobar-consumer-group"
+    seedTopicWithRecords(topic1, 1000)
+
+    val producer = transactionalProducer(transactionalId)
+
+    var consumer = transactionalConsumer(consumerGroupId)
+    consumer.subscribe(List(topic1))
+    producer.initTransactions()
+
+    val random = new Random()
+    var recordsProcessed = 0
+    try {
+      while (recordsProcessed < 1000) {
+        producer.beginTransaction()
+        val shouldCommit = if (random.nextInt() % 2 == 0) true else false
+
+        val records = pollUntilAtLeastNumRecords(consumer, Math.min(10, 1000 - recordsProcessed))
```

Comment:
Seems like there's a good chance that all 1000 records will be returned on the first poll. Perhaps we can just divide the commit and abort cases into two separate tests so that we're sure we hit them both?